posts/inside-rust/2023-11-10-parallel-rustc.md

```diff
@@ -0,0 +1,239 @@
+read-write locks, atomic types are used where appropriate, and many front-end
+operations are made parallel. The addition of parallelism was done by modifying
+a relatively small number of key points in the code. The vast majority of the
+front-end code did not need to be changed.
+
+When the parallel front-end is enabled and configured to use eight threads, we
+get the following Samply profile when compiling the same example as before.
+
+![Samply output when compiling Cargo, parallel](../../../../images/inside-rust/2023-11-10-parallel-rustc/samply-parallel.png)
+
+Again, there are several things worth nothing.
+- Front-end execution takes 5.9 seconds (down from 10.2 seconds).
+- Back-end execution takes 5.3 seconds (down from 6.2 seconds), and the LLVM
+  threads are running for 4.9 seconds of that (down from 5.9 seconds).
+- There are seven additional threads labelled "rustc" operating in the
+  front-end. The reduced front-end time shows they are reasonably effective,
+  but the thread utilization is patchy, with the eight threads all having
+  periods of inactivity. There is room for significant improvement here.
+- Eight of the LLVM threads start at the same time. This is because the eight
+  "rustc" threads create the LLVM IR for eight codegen units in parallel. (For
+  seven of those threads that is the only work they do in the back-end.) After
+  that, the staircase effect returns because only one "rustc" thread does LLVM
+  IR generation while seven or more other LLVM threads are active. If the
+  number of threads used by the front-end was changed to 16 the staircase shape
+  would disappear entirely, though in this case the final execution time would
+  barely change.
+
+### Putting it all together
+
+Rust compilation has long benefited from intraprocess parallelism, via Cargo,
+and from interprocess parallelism in the back-end. It can now also benefit from
+interprocess parallelism in the front-end.
+
+Attentive readers might be wondering how intraprocess parallelism and
+interprocess parallelism interact. If we have 20 parallel rustc invocations and
+each one can have up to 16 threads running, might we end up 100s of threads on
+a machine with only 10s of cores, resulting in inefficient execution as the OS
+tries its best to schedule them?
+
+Fortunately no. The compiler uses the [jobserver
+protocol](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html)
+to limit the number of threads it creates. If a lot of interprocess parallelism
+is occuring, intraprocess parallelism will be limited appropriately, and
+the number of threads will not exceed the number of cores.
+
+## How to use it
+
+As of (XXX: date enabled) the nightly compiler is [shipped with the parallel
+front-end enabled](https://github.com/rust-lang/rust/pull/117435). However,
+**by default it runs in single-threaded mode** and won't reduce compile times.
+Keen users who want to try multi-threaded mode should use the `-Z threads=8`
```

Comment:
I might mention that `-Z threads=n` where `n` is above the physical core count would not be recommended.

Reply:
Maybe not. On my 48-core machine, the performance of 16 threads is not even as good as 8 threads. And it can be observed that the utilization rate of many threads is low.

Reply:
Based on these comments, I think the current text is ok.